merkle-tree/indexed/src/lib.rs

```diff
@@ -171,58 +217,85 @@ where
-            .position(|element| element.index == low_element.index);
-        // key (index) value index in the changelog
-
-        match changelog_element_index {
-            Some(changelog_element_index) => {
-                let max_usize = usize::MAX;
-                // TODO: benchmark whether overwriting or the comparison is more expensive.
-                // Removed elements must not be used again.
-                if changelog_element_index == max_usize {
-                    return Err(IndexedMerkleTreeError::LowElementNotFound);
+            .iter_from(indexed_changelog_index)?
+            .skip(1)
+            .enumerate()
+            .filter_map(|(index, changelog_entry)| {
+                if changelog_entry.element.index == low_element.index {
+                    Some(indexed_changelog_index + 1 + index)
+                } else {
+                    None
                 }
-                let changelog_element = &mut self.indexed_changelog[changelog_element_index];
-                let patched_element = IndexedElement::<I> {
-                    value: BigUint::from_bytes_be(&changelog_element.value),
-                    index: changelog_element.index,
-                    next_index: changelog_element.next_index,
-                };
-                // Removing the value:
-                // Writing data costs CU thus we just overwrite the index
-                // with an impossible value so that it cannot be found.
-                changelog_element.index = max_usize
-                    .try_into()
-                    .map_err(|_| IndexedMerkleTreeError::IntegerOverflow)?;
-                // Only use changelog event values, since these originate from an account -> can be trusted
-                Ok(Some((patched_element, changelog_element.next_value)))
-            }
-            None => Ok(None),
+            })
+            .collect();
+
+        for changelog_element_index in changelog_element_indices {
+            let changelog_entry = &mut self.indexed_changelog[changelog_element_index];
+
+            // Patch the changelog index.
+            *changelog_index = changelog_entry.changelog_index;
+
+            // Patch the `next_index` of `new_element`.
+            new_element.next_index = changelog_entry.element.next_index;
+
+            // Patch the element.
+            low_element.update_from_raw_element(&changelog_entry.element);
+            // Patch the next value.
+            *low_element_next_value = BigUint::from_bytes_be(&changelog_entry.element.next_value);
+            // Patch the proof.
+            low_leaf_proof.clone_from(&changelog_entry.proof);
+
+            // Removing the value:
+            // Writing data costs CU thus we just overwrite the index
+            // with an impossible value so that it cannot be found.
+            changelog_entry.element.index = usize::MAX
+                .try_into()
+                .map_err(|_| IndexedMerkleTreeError::IntegerOverflow)?;
```

Comment:
just copying that here fore visibility:


Idea for performance improvement:
1. just copy the data when you are sure that you will return it, in particular for the proof
2. iterate over the filtered changelogs in reverse
    2.1 if the first element (the last filtered changelog element is still the low element we can stop there)
    2.2 if not we need to search element where it changes, **then make sure that the new low element is still the low element now** (this part might not yet be covered in ``patch_low_element_and_proof`` is it?)

Reply:
Going to try these ideas after I'm fully done with tests - I think they're going to be good with catching potential regressions or mistakes.